hoplite-core/src/main/kotlin/com/sksamuel/hoplite/ConfigSource.kt

```diff
@@ -11,6 +11,10 @@ import java.nio.file.Path
 
 abstract class ConfigSource {
 
+  /**
+   * Opens a new [InputStream], the caller should call [AutoCloseable.close]
+   * when they no longer need the [InputStream].
+   */
   abstract fun open(): ConfigResult<InputStream>
```

Comment:
This doc doesn't match if we're using use, since that close for you ?

Reply:
Here we are only calling `.use { }` on the result of `.open()`

`use { ... }` calls `close()` is a safe way after the code block is executed. 
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

> Executes the given block function on this resource and then closes it down correctly whether an exception is thrown or not.

General pattern is `inputStreamCreationFunction().use { ... }` which makes sure the `Closable` (i.e. inputstream) doesn't get used after close is called, but also ensure `close()` is called in a safe and reliable way.

similar to try-with-resources in Java, this helps avoid the issue mentions in Effective Java 3rd Ed, Item 9.

Reply:
Yes, but your javadoc says the caller should call [AutoCloseable.close] but by using `use` you're calling close for them. right ?

Reply:
yes. 

The caller of `open` needs to also ensure that `close` gets called. 

Is this case, we are calling `open` from the `ConfigFilePropertySource.node` function, so we also need to call `use`. 

We **don't** call `use` (or `close`) from _within_ the `open` function. 

Do you have a proposed change to the JavaDoc that would help make this clearer?

Reply:
Ah I see, I was looking at javadoc thinking it was applying to the concrete implementation. Thanks for the clarification.